gensim/models/lsimodel.py

```diff
@@ -489,7 +489,8 @@ def add_documents(self, corpus, chunksize=None, decay=None):
             chunksize = self.chunksize
         if decay is None:
             decay = self.decay
-
+        if not corpus:
```

Comment:
So I initially suggested `if corpus == []`, but in a previous revision you suggested `if not corpus` - personally I still think the first option is the safest, but if you have any other ideas I'm open :)

Reply:
Hmm, I probably didn't realize a `scipy.sparse` path was possible :)

But `corpus == []` would not work either – that will fail to detect empty generators, empty scipy.sparse arrays etc.

Reply:
Blah, that will silently eat away one document if `corpus` is a generator. Not good, scratch that.

Checking for emptiness in a general case sounds pretty complicated. How about we allow some false positives:

```
def is_empty(corpus):
    """Is the corpus (an iterable or a scipy.sparse array) empty?"""
    if scipy.sparse.issparse(corpus):
       return corpus.shape[1] > 0  # by convention, scipy.sparse documents are columns
    if isinstance(gen, types.GeneratorType):
       return False  # don't try to guess the content of generators, to avoid reading irretrievable elements
    try:
        first_doc = next(iter(corpus))  # list, numpy array etc
        return True  # first document exists => not empty
    except Exception:
        return False
```

Reply:
This looks ok, but something just occurred to me - does `__len__` have to be defined for any corpus? If not, maybe this would be a good requirement for a later version, as I feel like this would be useful in many places.

If it's defined we can just say:
```py
if len(corpus) == 0:
````
which works certainly for numpy, scipy.sparse I presume, and lists
...also if len has to be defined for generators and iterators

If not, I suppose something like this:
```py
def is_empty(corpus):
     try:
            length = len(corpus)
     except TypeError:
            return False
     return length == 0
```

Reply:
Also, I think the bools in your example are the wrong way round, perhaps this:


```py
def is_empty(corpus):
    """Is the corpus (an iterable or a scipy.sparse array) empty?"""
    if scipy.sparse.issparse(corpus):
       return corpus.shape[1] > 0  # by convention, scipy.sparse documents are columns
    if isinstance(gen, types.GeneratorType):
       return False  # don't try to guess emptiness of generators, may lose elements irretrievably
    try:
        first_doc = next(iter(corpus))  # list, numpy array etc
        return False # first document exists => not empty
    except Exception:
        return True
```

Reply:
No, `len` doesn't work on scipy.sparse, nor on generators.

Reply:
>  No, len doesn't work on scipy.sparse, nor on iterables in general.

Ah shame about scipy.sparse - in terms of generators in general, this is true - I suppose I meant something like this:
```py
from typing import TypeVar, Generator, Generic
YieldType = TypeVar("YieldType")
SendType = TypeVar("SendType")
ReturnType = TypeVar("ReturnType")

class AbstractCorpusGenerator(Generic[YieldType, SendType, ReturnType], Generator[YieldType, SendType, ReturnType]):
      def __len__(self) -> int:
             raise NotImplementedError
```
...but this is well outside the scope of this PR - just a little food for thought :)

Reply:
> No, `len` doesn't work on scipy.sparse, nor on iterables in general.
> 
> > Also, I think the bools in your example are the wrong way round, perhaps this:
> 
> You're right – needs to be switched, in multiple places:
> 
> ```python
> def is_empty(corpus):
>     """Is the corpus (an iterable or a scipy.sparse array) empty?"""
>     if scipy.sparse.issparse(corpus):
>        return corpus.shape[1] == 0  # by convention, scipy.sparse documents are columns
>     if isinstance(gen, types.GeneratorType):
>        return False  # don't try to guess emptiness of generators, may lose elements irretrievably
>     try:
>         first_doc = next(iter(corpus))  # list, numpy array etc
>         return False # first document exists => not empty
>     except Exception:
>         return True
> ```

Also yeah good spot on the other inequality - I think we can go with this version then if you're happy?

Reply:
Oh one final tweak
```py
def is_empty(corpus):
    """Is the corpus (an iterable or a scipy.sparse array) empty?"""
    if scipy.sparse.issparse(corpus):
       return corpus.shape[1] == 0  # by convention, scipy.sparse documents are columns
    if isinstance(gen, types.GeneratorType):
       return False  # don't try to guess emptiness of generators, may lose elements irretrievably
    try:
        next(iter(corpus))  # list, numpy array etc
        return False # first document exists => not empty
    except Exception:
        return True
```

Reply:
Nah, the original was more explicit = self-documenting = clearer.

Reply:
Eh I guess - normally I would avoid assigning variables that aren't used, although the performance is minimal I suppose. 

Also, just realised, any other error than stop iteration (e.g. the corpus cannot be made iterable) should return False - only "StopIteration" tells us it's an empty iterator:

```py
def is_empty(corpus):
    """Is the corpus (an iterable or a scipy.sparse array) empty?"""
    if scipy.sparse.issparse(corpus):
       return corpus.shape[1] == 0  # by convention, scipy.sparse documents are columns
    if isinstance(gen, types.GeneratorType):
       return False  # don't try to guess emptiness of generators, may lose elements irretrievably
    try:
        first_doc = next(iter(corpus))  # list, numpy array etc
        return False # first document exists => not empty
    except StopIteration:
        return True
    except Exception:
        return False
 ```

Reply:
You're right. Much better (safer) this way :)

Reply:
Ok cool - I'll add this to the PR :)